internal/alb/sg/association.go

```diff
@@ -0,0 +1,266 @@
+package sg
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/ingress/controller/store"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/alb/tg"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/aws/albec2"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/aws/albelbv2"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/pkg/util/log"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/pkg/util/types"
+)
+
+// Association represents the desired state of securityGroups & attachments for an Ingress resource.
+type Association struct {
+	// We identify Association by LbID
+	LbID string
+
+	LbArn          string
+	LbPorts        []int64
+	LbInboundCIDRs types.Cidrs
+
+	// ExternalSGIDs are custom securityGroups intended to be attached to LoadBalancer.
+	// If customers specified these securityGroups via annotation on ingress, the ingress controller will then stop creating securityGroups for loadbalancer or ec2-instances.
+	ExternalSGIDs []string
+
+	Targets tg.TargetGroups
+}
+
+// AssociationController provides functionality to manage Association
+type AssociationController interface {
+	// Reconcile ensured the securityGroups in AWS matches the state specified by assocation.
+	Reconcile(*Association) error
+	
+	// Delete ensures the securityGroups created by ingress controller for specified LbID doesn't exists.
+	Delete(*Association) error
+}
+
+// NewAssociationController constructs a new association controller
+func NewAssociationController(store store.Storer, ec2 *albec2.EC2, elbv2 albelbv2.ELBV2API) AssociationController {
+	logger := log.New("sg.association")
```

Comment:
gotta figure out some way to use a logger that has the ingress id instead of `sg.association`

Reply:
maybe pass the ingress name into the object been reconciled.
anyway, will do it in separate PR